### PR TITLE
1.- Signup signin

### PR DESCRIPTION
**Pull request by pfariaz**
# Summary

- User database model with migrations
- new endpoint for sign up users (POST /users)
- new endpint for sign in users (POST /users/sessions
- test cases for the 2 new endpoints

### CODE REVIEW

File: app/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  if (!regexValidEmail.test(userLogin.email)) {
+    return next(errors.defaultError('it must be a valid wolox.com email'));
+  }
+
+  userService
+    .getByEmail(userLogin.email)
+    .then(user => {
+      if (user != null) {
```
Comment (hdf1996) [inline app/controllers/users.js]: Try to always use three equals symbols instead of one (in the case of the negation there are just two signs), should be !==

Comment (hdf1996) [reply app/controllers/users.js]: If it's the first time you see this different aproach, take a look at https://stackoverflow.com/questions/523643/difference-between-and-in-javascript

File: app/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        bcrypt.compare(userLogin.password, user.password).then(valid => {
+          if (valid) {
```
Comment (hdf1996) [inline app/controllers/users.js]: Can't we make something like:
```
if(!valid) return next(errors.invalidUser());
// Rest of the code
```
? Personally i prefer to do an early return every time it's possible

File: app/models/scripts/dataCreation.js [hunk 3]
```diff
@@ -1,3 +1,50 @@
...
+          firstName: 'persona3',
+          lastName: 'apellido3',
+          email: 'email3@wolox.com',
+          password: hash
+        })
+      );
+
+      inserts.push(
+        db.models.user.create({
+          firstName: 'persona4',
```
Comment (hdf1996) [inline app/models/scripts/dataCreation.js]: You can use https://github.com/marak/Faker.js/ here, most of the time is great to have random data here (maybe you can keep the email and the password, and just use this in the name

File: app/orm.js [hunk 4]
```diff
@@ -4,9 +4,10 @@ const Sequelize = require('sequelize'),
 
 exports.DB_URL =
   config.common.database.url ||
-  `postgres://${config.common.database.username}
-  :${config.common.database.password}@${config.common.database.host}
-  :${config.common.database.port}/${config.common.database.name}`; // eslint-disable-line max-len
+  `postgres://${config.common.database.username}:${config.common.database.password}
+  @${config.common.database.host}:${config.common.database.port}/${config.common.database.name}`; // eslint-disable-line max-len
+
+console.log(exports.DB_URL);
```
Comment (hdf1996) [inline app/orm.js]: Do you need this? You are exposing your database password here

File: test/users.js [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+  should = chai.should();
+
+describe('users', () => {
+  describe('/users POST', () => {
+    it('should be success', done => {
+      chai
+        .request(server)
+        .post('/users')
+        .send({
+          firstName: 'persona5',
```
Comment (hdf1996) [inline test/users.js]: Remember faker? Here would be another good place to use it

File: test/users.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+          dictum.chai(res);
+        })
+        .then(() => done());
+    });
+    it('should fail because email exist', done => {
+      chai
+        .request(server)
+        .post('/users')
+        .send({
+          firstName: 'persona6',
```
Comment (hdf1996) [inline test/users.js]: idem here

File: test/users.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        .request(server)
+        .post('/users')
+        .send({
+          firstName: 'persona5',
+          lastName: 'apellido5',
+          password: 'falabella2017',
+          email: 'email5@wolox.com'
+        })
+        .then(res => {
+          res.should.have.status(200);
```
Comment (hdf1996) [inline test/users.js]: It's possible to check if the user has been created?

File: test/users.js [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        .post('/users')
+        .send({
+          firstName: 'persona6',
+          lastName: 'apellido6',
+          password: 'falabella2017',
+          email: 'email5@wolox.com'
+        })
+        .catch(err => {
+          err.should.have.status(400);
+          err.response.should.be.json;
```
Comment (hdf1996) [inline test/users.js]: idem here (with adding the test to check if the user was created), but with the negative case

File: test/users.js [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+          err.response.body.should.have.property('error');
+        })
+        .then(() => done());
+    });
+    it('should fail because password requirements', done => {
+      chai
+        .request(server)
+        .post('/users')
+        .send({
+          firstName: 'persona7',
```
Comment (hdf1996) [inline test/users.js]: Faker again

File: test/users.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+          lastName: 'apellido7',
+          password: 'fala1',
+          email: 'email7@wolox.com'
+        })
+        .catch(err => {
+          err.should.have.status(400);
```
Comment (hdf1996) [inline test/users.js:61]: Check if the user was created or not

File: test/users.js [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        })
+        .then(() => done());
+    });
+    it('should fail because required parameters are missing', done => {
+      chai
+        .request(server)
+        .post('/users')
+        .send({
+          firstName: 'persona8',
+          password: 'falabella2017',
```
Comment (hdf1996) [inline test/users.js]: Faker

File: test/users.js [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+          email: 'email8@wolox.com'
+        })
+        .catch(err => {
+          err.should.have.status(400);
+          err.response.should.be.json;
```
Comment (hdf1996) [inline test/users.js:80]: Check if the user was created or not

**Review by hdf1996 — CHANGES_REQUESTED** [review]

**Comment by pfariaz** [comment]
@hdf1986 now i fixed all the requested changes, thanks for the comments and tips :)

**Review by hdf1996 — APPROVED** [review]
Good job! Go on and rebase the next PR (Take care, with aaalll these changes it will be probably hard)
